Doc/library/functions.rst

```diff
@@ -1720,50 +1720,84 @@ are always available.  They are listed here in alphabetical order.
-   the *i*-th element from each of the argument sequences or iterables.  The
-   iterator stops when the shortest input iterable is exhausted. With a single
-   iterable argument, it returns an iterator of 1-tuples.  With no arguments,
-   it returns an empty iterator.  Equivalent to::
-
-        def zip(*iterables):
-            # zip('ABCD', 'xy') --> Ax By
-            sentinel = object()
-            iterators = [iter(it) for it in iterables]
-            while iterators:
-                result = []
-                for it in iterators:
-                    elem = next(it, sentinel)
-                    if elem is sentinel:
-                        return
-                    result.append(elem)
-                yield tuple(result)
-
-   The left-to-right evaluation order of the iterables is guaranteed. This
-   makes possible an idiom for clustering a data series into n-length groups
-   using ``zip(*[iter(s)]*n)``.  This repeats the *same* iterator ``n`` times
-   so that each output tuple has the result of ``n`` calls to the iterator.
-   This has the effect of dividing the input into n-length chunks.
-
-   :func:`zip` should only be used with unequal length inputs when you don't
-   care about trailing, unmatched values from the longer iterables.  If those
-   values are important, use :func:`itertools.zip_longest` instead.
-
-   :func:`zip` in conjunction with the ``*`` operator can be used to unzip a
-   list::
-
-      >>> x = [1, 2, 3]
-      >>> y = [4, 5, 6]
-      >>> zipped = zip(x, y)
-      >>> list(zipped)
-      [(1, 4), (2, 5), (3, 6)]
-      >>> x2, y2 = zip(*zip(x, y))
-      >>> x == list(x2) and y == list(y2)
-      True
+.. function:: zip(*iterables, strict=False)
+
+   Invert a series of iterables, turning rows into columns, and columns into rows.
+
+   Example::
+
+      >>> for item in zip(('a', 'b', 'c'), (1, 2, 3)):
+      ...     print(item)
+      ...
+      ('a', 1)
+      ('b', 2)
+      ('c', 3)
+
+   More formally: :func:`zip` returns an iterator of tuples, where the *i*-th
+   tuple contains the *i*-th element from each of the argument sequences or
+   iterables.
+
+   :func:`zip` returns a lazy iterable. The elements wouldn't be processed
+   until the iterable is iterated on, e.g. by wrapping in a :class:`tuple`.
+
+   One possible wrinkle is that the iterables passed to :func:`zip` could have
+   different lengths; sometimes by design, and sometimes because of a bug in
+   the code that prepared these iterables.  Python offers three different
+   approaches to dealing with this issue:
+
+   * By default, :func:`zip` has short-circuit behavior.  If any of the
+     iterables are shorter than the rest, :func:`zip` will ignore the remainders
+     of the longer iterables, cutting off the result to the length of the shortest
+     iterable::
+
+        >>> tuple(zip(range(3), ['fee', 'fi', 'fo', 'fum']))
+        ((0, 'fee'), (1, 'fi'), (2, 'fo'))
```

Comment:
I'd use `list(...)`. Also show the other case, `zip(range(10), ['fee', 'fi', 'fo', fum'])` (to clarify by example that it's not the length of the first iterable that determines the outcome).

Reply:
I've changed it to `list`, but I disagree about showing the other case. The entire topic of different lengths of iterables is by itself pretty esoteric, and it already gets a lot of words devoted to it in the `zip` documentation, so I want to avoid adding too much. If you feel strongly about this, I'm willing to let it go and include the other case.

Reply:
Okay, that's fair. Just don't claim it's esoteric on python-ideas. :-)